### PR TITLE
upgrade go mod

### PR DESCRIPTION
**Pull request by ruanxin**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,14 +1,20 @@
 module github.com/kyma-incubator/reconciler
 
-go 1.18
+go 1.21
+
+replace github.com/imdario/mergo => github.com/imdario/mergo v0.3.5
+
+replace github.com/docker/docker => github.com/docker/docker v23.0.7-0.20230730020554-801e90549aac+incompatible
```
Comment (ruanxin) [inline go.mod:7]: To reviewer:
have to use this verison atm https://github.com/testcontainers/testcontainers-go/issues/1359#issuecomment-1657098312

**Comment by ruanxin** [comment]
`pre-main-reconciler-publish-pr-cli` failing is expected, because the dependencies upgraded in this PR is not compatible with current cli, after this get merged, in CLI need to upgrade reconciler dependencies.

**Review by Tomasz-Smelcerz-SAP — APPROVED** [review]

**Pull request merged**